src/DynamoUtilities/SmartObservableCollection.cs

```diff
@@ -24,15 +27,25 @@ public class SmartObservableCollection<T> : ObservableCollection<T>
         /// </summary>
         internal event NotifyCollectionChangedEventHandler CollectionChangedDuringDeferredReset;
 
-        public SmartObservableCollection() : base()
+        public SmartObservableCollection() : this(null)
         {}
 
-        public SmartObservableCollection(List<T> list)
-            : base(list)
-        {}
-
-        public SmartObservableCollection(IEnumerable<T> collection) : base(collection)
+        public SmartObservableCollection(List<T> list) : this(list as IEnumerable<T>)
         {}
+  
+        public SmartObservableCollection(IEnumerable<T> collection) : base()
+        {
+            try
+            {
+                //Replace internal IList (from Collection<T>.items) with a thread safe IList(ThreadSafeList)
+                var itemsField = typeof(Collection<T>).GetField("items", BindingFlags.NonPublic | BindingFlags.Instance);
+                itemsField.SetValue(this, new ThreadSafeList<T>(items: collection));
```

Comment:
How expensive is this?

Reply:
Instead of doing this manipulation, cannot we derive this smart class from a thread-safe observable collection of some sort, or does that not exist?

Reply:
I could not find a thread-safe observable collection out of the box.
And the initial direction I took was to keep the SmartObservableCollection class compatible with (derived from) the ObservableCollection class (for API compatibility)

Reply:
This reflection code is only called when the class is constructed

Reply:
I had some more comments on this matter in the Conversation tab..but I will copy it here

If this seems like a bit of a hack...that's because it is (there is no protected or public way to change the inner field `items`)

The purpose of doing this forced `set` is to allow SmartObservableCollection instances to be thread safe even when passed around as base class variables.
Ex. 
ObservableCollection obs = new SmartObservableCollection();
Collection obs = new SmartObservableCollection();
In both cases interacting with obs will be thread safe.

All I am doing is replacing the internal (private) IList items that Collection is using under the hood with a ThreaSafeList instance. The items field is only accessed as an IList...so passing another implementation of IList will work ok
Here is the source code for Collection https://referencesource.microsoft.com/#mscorlib/system/collections/objectmodel/collection.cs,281923b8611114ec

The only risk I see...is if Microsoft decides to change their internal workings and maybe rename or remove the private field Collection<T>.items
In that case we now throw an exception....or we could leave the SmartObservableCollection to be thread-unsafe...
 
Alternatives that I can think of:
1. Lose API compatibility with the ObservableCollection and Collection classes and implement(or re-use) an entirely different thread safe collection class
2. Keep API compatibility with the ObservableCollection and Collection classes but loose thread safety when passed around as those types of class variables
3. Give up on thread safety at this level.

Reply:
This makes me wonder, why this does not exist already, and if we are in for some quite difficult to debug WPF threading interactions with bindings.

Reply:
class Collection<T> can handle passing  custom IList implementations 
```
public Collection(IList<T> list) {
            if (list == null) {
                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
            }
            items = list;
        }
```
ObservableCollection<T> : Collection<T> always tries to create a new list that it then passes to the base class (probably to avoid referencing other ObservableCollections ...)
```
        public ObservableCollection(List<T> list)
            : base((list != null) ? new List<T>(list.Count) : list)
        {
            // Workaround for VSWhidbey bug 562681 (tracked by Windows bug 1369339).
            // We should be able to simply call the base(list) ctor.  But Collection<T>
            // doesn't copy the list (contrary to the documentation) - it uses the
            // list directly as its storage.  So we do the copying here.
            // 
            CopyFrom(list);
        }
```
All list management APIs from ObservableCollections are defered to its base class Collection...
All this said...I do not think we will have issues with `difficult to debug WPF threading interactions with bindings`.

Reply:
Reverted change